### PR TITLE
Karenchen/enable draftv2

### PR DESCRIPTION
**Pull request by karenychen**
---
This is an extension for DraftV2. We have created a new feature branch as Marcus is away for the rest of the week
[DraftV2 Repo](https://github.com/Azure/draftv2)
[DraftV2 Design Dock](https://msazure.visualstudio.com/CloudNativeCompute/_wiki/wikis/CloudNativeCompute.wiki/226132/az-aks-app-Design-Doc)

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [x] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your PR is merged into master branch, a new PR will be created to update `src/index.json` automatically.  
The precondition is to put your code inside this repo and upgrade the version in the PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
AKS

File: src/aks-preview/azext_aks_preview/aks_app_commands/aksappcommands.py
Comment (Tatsinnit) [inline src/aks-preview/azext_aks_preview/aks_app_commands/aksappcommands.py:51]: 💡 This is `amd64` arch based so is this only going to work for some OS and not all?

File: src/aks-preview/azext_aks_preview/aks_app_commands/aksappcommands.py
Comment (Tatsinnit) [inline src/aks-preview/azext_aks_preview/aks_app_commands/aksappcommands.py:30]: 💡 📓 Quick note/observation: If not successful then I hope we are bubbling the error up.

File: src/aks-preview/azext_aks_preview/aks_app_commands/aksappcommands.py
Comment (Tatsinnit) [inline src/aks-preview/azext_aks_preview/aks_app_commands/aksappcommands.py:125]: 💡 Idea/food for thought: This seems like the download is depended on a release version. So imagine scenario that if for example version `0.0.v7` fixes something key, the deployment for correct fix has to happen separately. If we use `latest` tag then no matter what the release will be as long as latest points to that release we will always download the latest.

This needs to be discuss with `draftv2` team and what their delivery/release strategy will be.

File: src/aks-preview/azext_aks_preview/aks_app_commands/aksappcommands.py
Comment (Tatsinnit) [inline src/aks-preview/azext_aks_preview/aks_app_commands/aksappcommands.py:130]: 💡 We could get inputs form existing `cli` devs to double check if they have any other preferred way for `download`

File: src/aks-preview/azext_aks_preview/aks_app_commands/aksappcommands.py
Comment (Tatsinnit) [inline src/aks-preview/azext_aks_preview/aks_app_commands/aksappcommands.py:146]: 🐛 🪲`Security issue` - This could cause a security issue, because whatever the binary is we are providing `read, write and execute` once downloaded we could do, `0o655` or whatever is preferred permissioning model.

**Review by Tatsinnit — COMMENTED** [review]
💡 Thank you so much for this, and looks great!! ☕️❤️🙏 I have added some initial comments, also you could rewrite the commits history and `silent` the list of changes from Marcus's repo.

Thank you, changes look good, please feel free to reach out whatever ways suite. I believe since `Draftv2` is not released yet, we can wait for the real tool to become live.

File: src/aks-preview/azext_aks_preview/aks_app_commands/aksappcommands.py
Comment (karenychen) [reply src/aks-preview/azext_aks_preview/aks_app_commands/aksappcommands.py:100]: @qpetraroia

File: src/aks-preview/azext_aks_preview/aks_app_commands/aksappcommands.py
Comment (imiller31) [reply src/aks-preview/azext_aks_preview/aks_app_commands/aksappcommands.py:125]: I think that latest is fine -- in the future we might pin an "az-cli" tag

File: src/aks-preview/azext_aks_preview/aks_app_commands/aksappcommands.py
Comment (Tatsinnit) [inline src/aks-preview/azext_aks_preview/aks_app_commands/aksappcommands.py:46]: 💡 Type checks will fail, please note, some code path returns `None` if that is the case then type annotation of the function should mention that for example: `def monty() -> str | None:`

* If keen more here: https://peps.python.org/pep-0483/  OR https://docs.python.org/3/library/typing.html

File: src/aks-preview/azext_aks_preview/aks_app_commands/aksappcommands.py
Comment (Tatsinnit) [inline src/aks-preview/azext_aks_preview/aks_app_commands/aksappcommands.py:107]: 💡 Just a note: this is empty function. (Just as fyi)

File: src/aks-preview/azext_aks_preview/aks_app_commands/aksappcommands.py
Comment (Tatsinnit) [inline src/aks-preview/azext_aks_preview/aks_app_commands/aksappcommands.py:112]: 💡 Please note: Multiple type annotation here as well.

**Review by Tatsinnit — APPROVED** [review]
☕️🙏:shipit: looks good, please don’t forget the _help.py changes and rest I guess we should wait for this to merge. Please make sure you have tested use cases locally. 

I am sure @FumingZhang and team might have some addition but looks good to me. 

+1 form my end. Thank you 🙏

File: src/aks-preview/azext_aks_preview/aks_app_commands/aksappcommands.py
Comment (FumingZhang) [inline src/aks-preview/azext_aks_preview/aks_app_commands/aksappcommands.py:38]: Consider using logger to output different levels of prompts.

Note that the content of print is output to stdout. With logger, the content above the warning level is output to stderr, and the content at the info level and below is not output by default.

**Review by FumingZhang — COMMENTED** [review]
Please resolve the conflicts and add a set of live test cases for this new command.

File: src/aks-preview/azext_aks_preview/aks_app_commands/aksappcommands.py
Comment (karenychen) [reply src/aks-preview/azext_aks_preview/aks_app_commands/aksappcommands.py:38]: Hey Fuming, can you link me to an example for writing live test cases that I can follow?

Comment (FumingZhang) [reply src/aks-preview/azext_aks_preview/aks_app_commands/aksappcommands.py:38]: live test cases in [aks-preview](https://github.com/Azure/azure-cli-extensions/blob/main/src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py), [azure-cli/acs](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py)

File: src/aks-preview/azext_aks_preview/aks_app_commands/aksappcommands.py
Comment (NickKeller) [inline src/aks-preview/azext_aks_preview/aks_app_commands/aksappcommands.py:140]: I think we want to download a specific version of aks-app here. That way you can control when customers get updates

Comment (karenychen) [reply src/aks-preview/azext_aks_preview/aks_app_commands/aksappcommands.py:140]: We had a discussion on this during the first iteration and the solution was to use the latest release for now and potentially change to a specific tag later. more info [here](https://github.com/Azure/azure-cli-extensions/pull/4651#discussion_r846028241)

File: src/aks-preview/azext_aks_preview/aks_app_commands/aksappcommands.py
Comment (NickKeller) [inline src/aks-preview/azext_aks_preview/aks_app_commands/aksappcommands.py:84]: I think you should just return a concatenation of `paths` and `result`. This will allow you to detect the draftv2 binary anywhere along the customer's PATH. Otherwise we're downloading the draftv2 binary for a second time, when it's already executable, since it's in one of the PATH directories

**Review by Marcus-Hines — APPROVED** [review]

**Comment by yonzhan** [comment]
AKS

**Comment by zhoxing-ms** [comment]
Could you add some tests for these new features?
If you want to release a new version to release these features, please update `HISTORY.rst` and `setup.py` at the same time

**Comment by zhoxing-ms** [comment]
In addition, is this PR still in the draft status? If so, please convert it to a draft PR

**Comment by Tatsinnit** [comment]
💡 Hope this helps: seems like there is a `cassette` file which is missing which is causing this I guess, here is the guidance for it: https://dev.azure.com/msazure/CloudNativeCompute/_wiki/wikis/CloudNativeCompute.wiki/156735/AZCLI-AKS-Live-Unit-Test-Pipeline?anchor=create-cassette-file(s)-for-newly-added-test-case(s) we could reach out to @FumingZhang if more help is in need. 

cc: @karenychen 🙏☕️

**Review by FumingZhang — COMMENTED** [review]
Queued a [pipeline](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=55660300&view=logs&j=b162b355-d59d-5864-ce0f-0a70f12dd28b&t=a26fe913-e2be-5062-b08e-d15f1acc7ea2&l=828) to test the newly added cases.

Found some problems with concurrent execution. These test cases all try to modify (download)/use the binary in the same location.

File: src/aks-preview/azext_aks_preview/aks_draft/commands.py
Comment (Tatsinnit) [inline src/aks-preview/azext_aks_preview/aks_draft/commands.py:175]: 💡Sharing a **quick observation\thoughts** : Please see line above: `draft_binary_path` so even if the binary will exist, we go ahead and download the binary? i.e. does exist makes any change to the decision tree of `download_binary` ?

Comment (qpetraroia) [reply src/aks-preview/azext_aks_preview/aks_draft/commands.py:175]: Thank you for this Tats!

File: src/aks-preview/azext_aks_preview/_help.py
Comment (zhoxing-ms) [inline src/aks-preview/azext_aks_preview/_help.py:1715]: Could you add some `examples` for these commands?

File: src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py
Comment (zhoxing-ms) [inline src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py:4202]: May I ask why can't these tests run in record mode?

Comment (karenychen) [reply src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py]: oops forgot to change it back when i was testing something :)

File: src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py
Comment (FumingZhang) [inline src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py:4178]: It's ok to check-in the code first if you have a strict timeline for this feature. But don't forget to add the test cases later.

In order to avoid concurrent execution errors, the command (or in the test case) should support downloading the required binary to a temporary directory instead of a uniform path.

Besides, if the case would download a binary, better mark it as @live_only, otherwise the recording file would include the binary (the size of [test_aks_draft_with_helm.yaml](https://github.com/Azure/azure-cli-extensions/pull/4651/files#diff-4d1c6742e67dcee8c5df01fec4041cd3dae3d180bc3eebf04387b0739619878d) is 653,572 ), which is unnecessary. The case would be executed periodically in cli runner to make sure the command is available.

Comment (qpetraroia) [reply src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py:4179]: @FumingZhang Once the test passes can you merge this?

**Comment by qpetraroia** [comment]
@FumingZhang ready for you to merge when you can 😀

**Comment by FumingZhang** [comment]
Queued another [pipeline](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=55687293&view=results) to verify test cases. All test cases passed.

**Review by FumingZhang — APPROVED** [review]
LGTM

**Review by zhoxing-ms — APPROVED** [review]

**Pull request merged**
